### PR TITLE
ActionHero.api globally available via require 

### PR DESCRIPTION
**Pull request by evantahler**
@crrobinson14 is a genius.  

Rather than pass the `api` object around all over the application, when it is needed, we can simply `const api = require ('actionhero').api`.  There are some tricky hoops to jump though regarding the definition of our base classes, but this creates a very pleasant developer API.

This changes the method signature of `run` for actions and tasks, initializers, servers, and CLI commands. 

## Action
```js
const {Action, api} = require('actionhero')

module.exports = class MyAction extends Action {
  constructor () {
    super()
    this.name = 'myAction'
    this.description = 'myAction'
    this.outputExample = {}
  }

  async run (data) {
    api.log('yay')
    // your logic here
  }
}
```

## Task
```js
const {Task, api} = require('actionhero')

module.exports = class MyTask extends Task {
  constructor () {
    super()
    this.name = 'myTask'
    this.description = 'myTask'
    this.frequency = 0
  }

  async run (data) {
    api.log('yay')
    // your logic here
  }
}
```

## Initializer

```js
const {Initializer, api} = require('actionhero')

module.exports = class MyInitializer extends Initializer {
  constructor () {
    super()
    this.name = 'myInitializer'
    this.loadPriority = 1000
    this.startPriority = 1000
    this.stopPriority = 1000
  }

  async initialize () {
    api['myInitializer'] = {}
  }

  async start () {}
  async stop () {}
}
```

## CLI Command

```js
const {CLI, api} = require('actionhero')

module.exports = class MyCLICommand extends CLI {
  constructor () {
    super()
    this.name = 'myCLI'
    this.description = 'myCLI'
    this.inputs = {}
  }

  async run (data) {
    api.log('yay')
    return true
  }
}
```

## Server
```js
const {Server, api} = require('actionhero')

module.exports = class MyServer extends Server {
  constructor () {
    super()
    this.type = 'myServer'

    this.attributes = {
      canChat: false,
      logConnections: true,
      logExits: true,
      sendWelcomeMessage: false,
      verbs: []
    }
    // this.config will be set to equal api.config.servers[this.type]
  }

  initialize () {
    this.on('connection', (conection) => {

    })

    this.on('actionComplete', (data) => {

    })
  }

  start () {
    // this.buildConnection (data)
    // this.processAction (connection)
    // this.processFile (connection)
  }

  stop () {

  }

  sendMessage (connection, message, messageCount) {

  }

  sendFile (connection, error, fileStream, mime, length, lastModified) {

  }

  goodbye (connection) {

  }
}
```

### CODE REVIEW

File: classes/process.js [hunk 1]
```diff
@@ -17,28 +22,25 @@ module.exports = class Process {
...
     }
 
-    this.api = {
-      running: false,
-      initialized: false,
-      shuttingDown: false,
-      projectRoot: projectRoot,
-      bootTime: null
-    }
+    api.running = false
```
Comment (gcoonrod) [inline classes/process.js]: Is there a reason these parameters aren't namespaced? e.g. `api.actionheroInstance.running`.

File: index.js [hunk 2]
```diff
@@ -24,3 +24,5 @@ const path = require('path');
 ].forEach(({klass, file}) => {
   exports[klass] = require(path.join(__dirname, 'classes', file))
 })
+
+exports.api = {}
```
Comment (gcoonrod) [inline index.js]: Each one of the classes in this file are going to require in `api` and then modify it. Why not create the `api` object at the start of the file and then run a reduce over the array of Classes and pass in `api`. That should remove the need for the weird `require` statements.

File: test/core/api.js [hunk 3]
```diff
@@ -65,25 +65,25 @@ describe('Core: API', () => {
...
           }
         },
         '2': {
           name: 'versionedAction',
           description: 'I am a test',
           version: 2,
           outputExample: {},
-          run: async (api, connection) => {
-            connection.response.version = 2
+          run: async (data) => {
```
Comment (gcoonrod) [inline test/core/api.js]: Why the change from `connection` to `data`?

File: actions/cacheTest.js [hunk 4]
```diff
@@ -50,7 +50,8 @@ module.exports = class CacheTest extends ActionHero.Action {
     }
   }
 
-  async run ({cache}, {params, response}) {
+  async run ({params, response}) {
+    const {cache} = ActionHero.api
```
[the remark this answers is not in the document]
Comment (evantahler) [reply actions/cacheTest.js]: Agree! I'll go with `const {Action, api} = require('./../index.js')`

File: bin/actionhero [hunk 5]
```diff
@@ -6,6 +6,7 @@ const path = require('path')
 const optimist = require('optimist')
 const spawn = require('child_process').spawn
 const ActionHero = require(path.join(__dirname, '..', 'index.js'))
+const api = ActionHero.api
```
[the remark this answers is not in the document]
Comment (evantahler) [reply bin/actionhero]: a little lower down we also do `const actionHeroProcess = new ActionHero.Process()`

File: bin/templates/cli.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 'use strict'
 const ActionHero = require('actionhero')
 
-module.exports = class MyCLICommand extends ActionHero.Action {
+module.exports = class MyCLICommand extends ActionHero.CLI {
```
[the remark this answers is not in the document]
Comment (evantahler) [reply bin/templates/cli.js]: 🤐

File: index.js (hunk 2 above)
Comment (evantahler) [reply index.js]: The 4 top-level properties of the api object are:
```
    api.running (bool)
    api.initialized (bool)
    api.shuttingDown (bool)
    api.projectRoot (string)
    api.bootTime (Date)
```

I really do think that these belong at the top of the stack (rather than under `api.status` or something, as the do describe the API itself.

File: test/core/api.js (hunk 3 above)
Comment (evantahler) [reply test/core/api.js]: `connection` was wrong, it's an acton so `data` is what you get.  I could have done `{response}` I guess, but I was just correcting a naming error.

File: classes/process.js (hunk 1 above)
Comment (evantahler) [reply classes/process.js]: The 5 top-level properties of the api object are:
```
    api.running (bool)
    api.initialized (bool)
    api.shuttingDown (bool)
    api.projectRoot (string)
    api.bootTime (Date)
```

I really do think that these belong at the top of the stack (rather than under `api.status` or something, as they do describe the API itself.

File: index.js (hunk 2 above)
Comment (evantahler) [reply index.js]: I tried a few methods of this, but since the `package.json#main` of `actionhero` is this index file, it has to be the one exporting... something.   If I have the `Process` return the API, that means I need to instantiate it within index, and I don't want to do that either.

Comment (evantahler) [reply index.js]: If you can come up with a better pattern, by all means please submit it! I'm *happy enough* with how it works now, as only the private classes are strange.  No user-facing files have the require problem.
